src/firebase-resource-manager.ts

```diff
@@ -0,0 +1,119 @@
+import * as clc from "cli-color";
+
+import * as api from "./api";
+import * as FirebaseError from "./error";
+import * as logger from "./logger";
+import { OperationPollerOptions, pollOperation } from "./operation-poller";
+import { OraWrapper } from "./oraWrapper";
+
+const ONE_SECOND_MILLIS = 1000;
+
+export enum ParentResourceType {
+  ORGANIZATION = "organization",
+  FOLDER = "folder",
+}
+
+export interface ParentResource {
+  id: string;
+  type: ParentResourceType;
+}
+
+export async function createFirebaseProject(
+  projectId: string,
+  projectDisplayName: string,
+  parentResource?: ParentResource
+): Promise<{ projectId: string }> {
+  await createCloudProject(projectId, projectDisplayName, parentResource);
+  const projectInfo = await addFirebaseToCloudProject(projectId);
+
+  logger.info("");
+  logger.info("🎉🎉🎉 Your Firebase project is ready! 🎉🎉🎉");
+  logger.info("");
+  logger.info("Project information:");
+  logger.info(`   - Project ID: ${clc.bold(projectInfo.projectId)}`);
+  logger.info(`   - Project Name: ${clc.bold(projectInfo.displayName)}`);
+  logger.info("");
+  logger.info("Firebase console is available at");
+  logger.info(`https://console.firebase.google.com/project/${clc.bold(projectId)}/overview`);
+  return { projectId };
+}
+
+/**
+ * Send an API request to create a new Google Cloud Platform project and poll the LRO to get the
+ * new project information.
+ * @return {Promise} this function returns a promise that resolves to the new cloud project
+ *     information
+ */
+async function createCloudProject(
+  projectId: string,
+  projectDisplayName: string,
+  parentResource?: ParentResource
+): Promise<any> {
+  const spinner = new OraWrapper("Creating Google Cloud Platform project");
+  spinner.start();
+
+  try {
+    const response = await api.request("POST", "/v1/projects", {
+      auth: true,
+      origin: api.resourceManagerOrigin,
+      timeout: 15 * ONE_SECOND_MILLIS,
+      data: { projectId, name: projectDisplayName, parent: parentResource },
+    });
+
+    const projectInfo = await pollOperation<any>({
+      pollerName: "Project Creation Poller",
+      apiOrigin: api.resourceManagerOrigin,
+      apiVersion: "v1",
+      operationResourceName: response.body.name /* LRO resource name */,
+    });
+    spinner.succeed();
```

Comment:
I don't think spinners should be placed inside the API-calling functions, as it too tightly couples output with behavior. Instead, the spinner should be placed in the command itself.

Reply:
I understand. However, the main reason I put spinners in API-calling functions is because I want the spinners to be more informative when the provisioning process includes multiple steps, so if the command fails, users know what has been done and at which step the command fails. In `projects:create` command, there are **create cloud project** step and **add Firebase to cloud project** step, each of which has a spinner. if the cloud project is created but the add Firebase step failed, user can see that the cloud project was created but not the Firebase project, thus they can not reuse the same `projectId` for the next calls.
An option is to export `createCloudProject()` and `addFirebaseToCloudProject()` functions instead of exporting a single function that includes both steps, that way I can have 2 spinners in the command file.

Reply:
Yes, I'd split it into two exported functions and move the spinner into the command.

Reply:
That works for me.

Reply:
Exported `createCloudProject()` and `addFirebaseToCloudProject()`, moved `createFirebaseProject()` to command file as a helper method.